master/internal/api_project.go

```diff
@@ -523,9 +524,29 @@ func (a *apiServer) PostProject(
 		return nil, status.Error(codes.PermissionDenied, err.Error())
 	}
 
+	var projectKey string
+	if req.Key == nil {
+		projectKey, err = project.GenerateProjectKey(ctx, req.Name)
+		if err != nil {
+			return nil, fmt.Errorf("error generating project key: %w", err)
+		}
+	} else {
+		projectKey = *req.Key
+	}
+
 	p := &projectv1.Project{}
 	err = a.m.db.QueryProto("insert_project", p, req.Name, req.Description,
-		req.WorkspaceId, curUser.ID)
+		req.WorkspaceId, curUser.ID, projectKey)
+
+	if err != nil && strings.Contains(err.Error(), db.CodeUniqueViolation) {
+		if strings.Contains(err.Error(), "projects_key_key") {
+			return nil,
+				status.Errorf(codes.AlreadyExists, "project with key %s already exists", projectKey)
+		} else if strings.Contains(err.Error(), "projects_name_key") {
+			return nil,
+				status.Errorf(codes.AlreadyExists, "project with name %s already exists", req.Name)
```

Comment:
Any reason why we can have 2 projects with the same name?

Reply:
My motivation for this was to return a 403 (already exists) status instead of returning an 500 since this wasn't an internal error rather an issue with the patch request.

Reply:
However, I'll remove it, this was more of a nit of my part for how we communicate errors. I think making sure that we are returning appropriate response codes is something we could improve.